### ISSUE TITLE
could not read from '/sys/module/pcc_cpufreq/initstate': No such device

### ISSUE THREAD

**Issue by cmurf** [systemd/systemd#20605]
**systemd version the issue has been seen with**
 > systemd-249.4-1.fc35.x86_64

**Used distribution**
 > Fedora 35

**Linux kernel version used** (`uname -a`)

 > `Linux 5.14.0-60.fc35.x86_64 #1 SMP Mon Aug 30 16:45:32 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > if it's not really a problem, then maybe it should be yellow text rather than red?

**Unexpected behaviour you saw**
 > [    2.923541] systemd-udevd[638]: could not read from '/sys/module/pcc_cpufreq/initstate': No such device

It's in red text so it's claiming this is urgent.

**Steps to reproduce the problem**
 > boot

**Additional program output to the terminal or log subsystem illustrating the issue**

```text
…
```

**Comment by cmurf** [comment]
[journal.txt](https://github.com/systemd/systemd/files/7088161/journal.txt)

**Comment by yuwata** [comment]
Hmm?? I cannot find the origin of the error message.

**Comment by yuwata** [comment]
Could you enable to generate debugging logs of udevd? Please try to boot with `udev.log_level=debug` kernel command line option.

**Comment by poettering** [comment]
It's libkmod that generates the log message:

https://github.com/lpereira/kmod/blob/master/libkmod/libkmod-module.c#L1729

I no idea what that means that the lib gets ENODEV when reading "initstate". Smells like a kernel problem?

@lucasdemarchi any idea?

**Comment by poettering** [comment]
(either way, not sure we can do anything about this, let's keep this open for discussion for now, but this appears to be an issue between the kernel and libkmod)

**Comment by lucasdemarchi** [comment]
@poettering built-in modules don't have a initstate file. Is pcc_cpufreq builtin or a module in Fedora 35?.

Regardless, it appears we could open the initstate file, but we couldn't read it. We do handle a race between the sysfs directory creation and the initstate creation (see https://github.com/lpereira/kmod/blob/master/libkmod/libkmod-module.c#L1753). However that race should not exist for reading the file, at least in theory. Did it start with a different systemd version or with a kernel upgrade?

**Comment by yuwata** [comment]
Just for the record, I also found the same log in my journal (udevd is built from git, kernel is 5.13.13-100.fc33.x86_64).
```
systemd-udevd[540]: LNXCPU:0a: Processing device (SEQNUM=2550, ACTION=add)
systemd-udevd[540]: LNXCPU:0a: /usr/lib/udev/rules.d/50-udev-default.rules:14 Importing properties from results of builtin command 'hwdb --subsystem=acpi'
systemd-udevd[540]: LNXCPU:0a: hwdb modalias key: "acpi:LNXCPU:"
systemd-udevd[540]: LNXCPU:0a: /usr/lib/udev/rules.d/80-drivers.rules:5 RUN 'kmod load $env{MODALIAS}'
systemd-udevd[540]: LNXCPU:0a: sd-device: Created db file '/run/udev/data/+acpi:LNXCPU:0a' for '/devices/LNXSYSTM:00/LNXCPU:0a'
systemd-udevd[540]: LNXCPU:0a: Running built-in command "kmod load acpi:LNXCPU:"
systemd-udevd[540]: Loading module: acpi:LNXCPU:
systemd-udevd[540]: Inserted module 'acpi_cpufreq'
systemd-udevd[540]: could not read from '/sys/module/pcc_cpufreq/initstate': No such device
systemd-udevd[540]: Inserted module 'pcc_cpufreq'
systemd-udevd[540]: LNXCPU:0a: Device processed (SEQNUM=2550, ACTION=add)
systemd-udevd[540]: LNXCPU:0a: sd-device-monitor: Passed 222 byte to netlink monitor
```

**Comment by poettering** [comment]
> @poettering built-in modules don't have a initstate file. Is pcc_cpufreq builtin or a module in Fedora 35?

It's not built in, it's a kernel module.

I wonder if https://lkml.org/lkml/2020/12/30/199 has anything to do with it.

**Comment by poettering** [comment]
Also note that the error is ENODEV, not ENOENT. The file appears to exist hence, but writing to it fails?

**Comment by gregkh** [comment]
no idea, sorry, I do not see how ENODEV is possible to returned from trying to read that sysfs file.

**Comment by poettering** [comment]
There are actually quite a bunch of reports about this, for various kmods:

https://www.google.com/search?q=%22No+such+device%22+initstate

**Comment by lucasdemarchi** [comment]
> Also note that the error is ENODEV, not ENOENT. The file appears to exist hence, but reading from it fails?

@poettering yes, that's what I meant by "Regardless, it appears we could open the initstate file, but we couldn't read it." - asking if it was a module was to understand what possible race we may having: the file existing for a fraction of time while the module is being loaded or something else.

@gregkh  wouldn't that be the case if the module failed to load and the kernfs file is removed?

Looking at my own machine, I do see this on older kernels:

```
[    1.852386] ldmartin-desk2 systemd[1]: Mounting /boot/efi...
[    1.863465] ldmartin-desk2 systemd-udevd[319]: could not read from '/sys/module/pcc_cpufreq/initstate': No such device
```

This was on Linux version 5.9.16 and was the only version in which I have this logged in journal. I never reproduced it on 5.13.x.

**Comment by cmurf** [comment]
Does seem to be a race, I'm not always seeing it. Same kernel and systemd version, 'no such device' does not appear, but multiple instances of 
>systemd-udevd[607]: Inserted module 'pcc_cpufreq'

[journaludevdebug.txt](https://github.com/systemd/systemd/files/7108081/journaludevdebug.txt)

**Comment by Zesko** [comment]
If I switched the kernel parameter `intel_pstate` to `amd-pstate` for AMD Ryzen,  I got this same error message randomly after booting.

Not every boot always shows the error message, but it appears probably 30% of all boots.

If I switch back to `intel_pstate`, then no error message anymore.
